migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs

```diff
@@ -29,6 +31,28 @@ impl PostgresFlavour {
 
 #[async_trait::async_trait]
 impl SqlFlavour for PostgresFlavour {
+    async fn acquire_lock(&self, connection: &Connection) -> ConnectorResult<()> {
+        // https://www.postgresql.org/docs/current/explicit-locking.html#ADVISORY-LOCKS
+
+        tokio::time::timeout(
+            ADVISORY_LOCK_TIMEOUT,
+            connection.raw_cmd("SELECT pg_advisory_lock(72707369)"),
+        )
+        .await
+        .map_err(|_elapsed| {
+            ConnectorError::user_facing_error(user_facing_errors::common::DatabaseTimeout {
+                database_host: connection.connection_info().host().to_owned(),
+                database_port: connection
+                    .connection_info()
+                    .port()
+                    .map(|port| port.to_string())
+                    .unwrap_or_else(|| "<unknown>".into()),
+            })
+        })??;
```

Comment:
I believe this code may cause a deadlock?

This attempts to get a lock. In user-land we configure a timeout. If we hit the timeout, the future is aborted -- but we never unset the request for the lock. If I'm not mistaken this may cause a situation where we acquire a lock, but never release it. Which is a deadlock.

I'd feel more comfortable if either:
1. this implemented the locking mechanism in SQL. This may be tricky, but [this thread](https://stackoverflow.com/questions/38853470/timeout-on-advisory-locks-in-postgresql) points towards it being possible.
2. we ran a destructor when the lock fails to be acquired.

I'm not super familiar with how this works though. If it's the case that this in fact can not deadlock, it'd be good to have a comment explaining how we're releasing the lock in the case of a timeout. This is hard to infer from the postgres docs alone, since we're interacting with it in interesting ways.

Reply:
I should document this in code, indeed, because it could easily be wrong in other circumstances. In our case, I think we don't have a problem, because assuming we time out, or fail in any other way, the error will bubble up to the user, and the CLI + migration engine processes will terminate, the connection will be ended and the lock will be released by the server. We don't have long lived connections, except when we are applying a migration and it takes a long time, in which case the lock is already acquired.